src/nvim/window.c

```diff
@@ -4019,24 +4019,25 @@ tabpage_T *win_find_tabpage(win_T *win)
   return NULL;
 }
 
-/*
- * Move to window above or below "count" times.
- */
-static void 
-win_goto_ver (
-    int up,                         /* TRUE to go to win above */
-    long count
-)
+/// Get the above or below neighbor window of the specified window.
+///
+/// Returns the specified window if the neighbor is not found.
+/// Returns the previous window if the specifiecied window is a floating window.
+///
+/// @param up     true for the above neighbor
+/// @param count  nth neighbor window
+///
+/// @return       found window
+win_T *win_vert_neighbor(tabpage_T *tp, win_T *wp, bool up, long count)
 {
   frame_T     *fr;
   frame_T     *nfr;
   frame_T     *foundfr;
 
-  foundfr = curwin->w_frame;
+  foundfr = wp->w_frame;
 
-  if (curwin->w_floating) {
-    win_goto(prevwin);
-    return;
+  if (wp->w_floating) {
+    return prevwin;
```

Comment:
Let me summarise it:

- Current behaviour:

  When the returned window is different from the current window, it could mean _two_ things: a window was found or the current window is a floating window and thus the previous window was returned.

- Maybe better behaviour:

  Only return a different window if there was an actual window found. A floating window will just return itself.

  So, a plugin author can use this:

  ```vim
  let right_win = winnr('l')
  let neighbor_found = right_win != winnr()
  ```